feat: bundle and push all journeys to kibana

vigneshshanmugam opened the pull request:
+ Part of #470 
+ Adds the subcommand `push` to the Synthetics CLI that would do the following
   - Goes over each journey file and bundles that particular journey along with relevant files that are imported by the particular journey
   - Creates a archived file for each journeys maintaining the local path (to help with Stack traces when invoked via HB)
   - creates a base 64 encoded content that would be sent to Kibana as part of the `MonitorSchema`. 
 + Introduce required options `--url` and `--auth` to the CLI flags that lets user specify the `Kibana URL endpoint` and also the `authentication` details. User auth accepts `username:pass` or `encoded` strings, If the auth parameter is not encoded, they would be encoded by the Synthetics agent before communicating with Kibana. 
+ In order to reduce the monitor id collision, it is required to set the project/suite id when invoking the push command. Users can pass it via `--project <name>`. 
+ Current schema looks like this for the create monitor part. 

```ts
type body = MonitorSchema[]

type MonitorSchema = {
  content: string; // base64 encoded monitor source
  id: string;
  name: string;
  schedule: string;
  locations: SyntheticsLocations[];
  throttling: {
    download: number;
    upload: number;
    latency: number;
  };
};

```

### To run locally

+ Install the deps and build the files by `npm run build`
```
npx @elastic/synthetics push examples/todos/basic.journey.ts --auth elastic:changeme  --url http://localhost:5601/wuv  --project test
```
+ Use the PR from Kibana - https://github.com/elastic/kibana/pull/131270 and do `yarn start`. Once Kibana is up use the kibana URL to push the monitors from local journey files.

apmmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Felastic-synthetics%2FPR-482/detail/PR-482/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Felastic-synthetics%2FPR-482/detail/PR-482/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Felastic-synthetics%2FPR-482/detail/PR-482/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Felastic-synthetics%2FPR-482/detail/PR-482/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://synthetics_482.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/apm-ci/transactions/view?rangeFrom=2022-03-31T22:20:03.042Z&rangeTo=2022-03-31T22:40:03.042Z&transactionName=apm-agent-rum/elastic-synthetics/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=dcd455c1c85167238138ea25aff47053&transactionId=12208e4ec26da8af)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-03-31T22:30:03.042+0000


* Duration: 14 min 1 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 149  |
| Skipped      | 2 |
| Total        | 151   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->












## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

vigneshshanmugam commented:
Notes from latest Refinement
* Purse the `monitor.use` API for the DSL 
    - We are fine with running the journey in `--dry-run` mode, It would gives us an idea in understanding the step names and a cleaner API for declarative syntax. 
    - We discussed the other option that would gives us the opportunity to figure out defaults without running Journeys, but decided against it. 
```ts
journey(
  { name: 'check if title is present', schedule: '@every1m' }, // First proposal
  ({ page, params, monitor }) => {
    monitor.use({.      // Second proposal
      id: 'blah',
      schedule: '@every1m',
    });
    step('launch app', async () => {
      await page.goto(params.url);
    });
  }
);

```

* Using Kibana auth for pushing monitors
   - Have a meeting scheduled with the Kibana Security team to understand the scope of work required to make the `API tokens` work, But worse case we would go with the `User credentials`.

vigneshshanmugam commented:
I have changed the way `monitor.use` API defined here and went exposing monitor object through the package as it offered better DX, extensibility and customization capabilities. Added more details on the PR - #493

vigneshshanmugam commented on src/cli.ts:
```diff
@@ -139,12 +140,20 @@ program
 
 program
-  .command('push')
+  .command('push [files...]')
```
It can do even multiple directories already, Example
```
npx @elastic/synthetics push ./examples/todos
```

vigneshshanmugam commented on src/push/index.ts:
```diff
@@ -0,0 +1,59 @@
+    const outPath = join(bundlePath, name + '.zip');
+    progress(`Preparing journey: ${name}`);
+    const value = await bundler.build(location.file, outPath);
+    console.log('Bundled', value);
```
Changed the structure a bit here. It was a throwaway code before.

vigneshshanmugam commented on src/push/plugin.ts:
```diff
@@ -0,0 +1,113 @@
+
+export function Plugin(callback: PluginCallback): esbuild.Plugin {
+  const isBare = (str: string) => {
+    if (str.startsWith('/') || str.startsWith('./') || str.startsWith('../')) {
+      return true;
+    }
```
Good point, I will add `node_modules` to the externals to which should catch this one. If the user is doing something with NPM Config prefix and fiddling with bundler, then its completely fine and on the HB side, we are not installing anything external other than the Synthetics library itself and the attack surface is also pretty limited on the service side.

vigneshshanmugam commented on src/push/bundler.ts:
```diff
@@ -0,0 +1,98 @@
+import path from 'path';
+import fs, { unlink } from 'fs/promises';
+import { createWriteStream } from 'fs';
+import * as esbuild from 'esbuild';
```
I will write up my thoughts soon given I have tried with almost all of the bundlers including Webpack, Esbuild, Rollup and Swc.

lucasfcosta reviewed (approved):
I've just E2E tested this against @dominiqueclarke's PR here: https://github.com/elastic/kibana/pull/131270. It worked fine, including the service.

However, I have a few pieces of feedback:

* When pushing a journey with the same ID to the same project, I get an unhelpful error message
    ![Screenshot 2022-05-18 at 14 16 35](https://user-images.githubusercontent.com/6868147/169047685-4c40621a-1a2f-4a45-96b4-265b49bc9007.png)
* I can still see a blank source field when trying to edit monitors
    ![Screenshot 2022-05-18 at 14 23 00](https://user-images.githubusercontent.com/6868147/169049042-766d0498-0a92-460d-bd51-7b1e47890d26.png)

vigneshshanmugam commented on src/push/monitor.ts:
```diff
@@ -0,0 +1,77 @@
+
+// Internal representation of Locations that would be used when
+// talking to the Kibana API
+const LocationsMap: Record<SyntheticsLocationsType, string> = {
+  'US East': 'us-east4-a',
+  'EU West': 'europe-west2-a',
+};
```
Yeah we talked about this in the last run on the refinement and decided to do patch release of Synthetics whenever we update/introduce new locations.

dominiqueclarke commented:
> npx @elastic/synthetics

@lucasfcosta I will make the monitor read-only in a separate PR.

vigneshshanmugam commented on src/cli.ts:
```diff
@@ -145,20 +147,38 @@ program
-    '--schedule <time>',
-    "The default interval for the pushed monitors. Setting `10m`, for example, configures monitors which don't have a specified interval defined to run every 10 minutes."
+    '--schedule <time-in-minutes>',
+    "schedule in minutes for the pushed monitors. Setting `10`, for example, configures monitors which don't have a specified interval defined to run every 10 minutes.",
```
Please suggest 👍🏽
